File: blackjack/ruby-fortran/partitions.f90
recursive integer function partitions(cards,subtotal) result(m)
  integer :: total
  integer, dimension(1:10) :: cards
  integer :: subtotal

  m = 0
  ! Hit
  do i = 1, 10
     if (cards(i)>0) then
        cards(i) = cards(i)-1
        total = subtotal+i
        if (total < 21) then
           ! Stand
           m = m+1
           ! Hit again
           m = m+partitions(cards, total)
        else if (subtotal+i==21) then
           ! Stand; hit again is an automatic bust
           m = m+1
        end if
        cards(i) = cards(i)+1
     end if
  end do

end function partitions

File: blackjack/ruby-nim/partitions.nim

proc partitions*(cards: var array[0..9, int], subtotal: int): int {. exportc, dynlib .} =
  var total: int
  result = 0

  # Hit
  for i in 0..9:
    if (cards[i]>0):
      total = subtotal+i+1
      if (total < 21):
        cards[i] -= 1
        # Stand
        result += 1
        # Hit again
        result += partitions(cards, total)
        cards[i] += 1
      elif (total==21):
        # Stand; hit again is an automatic bust
        result += 1
                
  return result
